webrender/src/device/gfx/device.rs

```diff
@@ -171,16 +172,54 @@ struct Fence<B: hal::Backend> {
 
 #[derive(Debug)]
 struct Frame<B: hal::Backend> {
-    image: ImageCore<B>,
-    depth: DepthBuffer<B>,
+    surface_image: Option<<B::Surface as PresentationSurface<B>>::SwapchainImage>,
     framebuffer: B::Framebuffer,
     framebuffer_depth: B::Framebuffer,
+    submitted: bool,
 }
 
 impl<B: hal::Backend> Frame<B> {
-    fn deinit(self, device: &B::Device, heaps: &mut Heaps<B>) {
-        self.image.deinit(device, heaps);
-        self.depth.deinit(device, heaps);
+    fn new(
+        device: &B::Device,
+        surface_image: <B::Surface as PresentationSurface<B>>::SwapchainImage,
+        depth: &B::ImageView,
+        viewport_rect: hal::pso::Rect,
+        render_passes: &HalRenderPasses<B>,
+    ) -> Self {
+
+        let image_extent = hal::image::Extent {
+            width: viewport_rect.w as _,
+            height: viewport_rect.h as _,
+            depth: 1,
+        };
+
+        let framebuffer = unsafe {
+            device.create_framebuffer(
+                &render_passes.bgra8,
+                Some(surface_image.borrow()),
+                image_extent,
+            )
+        }
+        .expect("create_framebuffer failed");
+
+        let framebuffer_depth = unsafe {
+            device.create_framebuffer(
+                &render_passes.bgra8_depth,
+                vec![surface_image.borrow(), &depth],
+                image_extent,
+            )
+        }
+        .expect("create_framebuffer failed");
+
+        Frame {
+            surface_image: Some(surface_image),
+            framebuffer,
+            framebuffer_depth,
+            submitted: false,
+        }
+    }
+
+    fn deinit(self, device: &B::Device) {
         unsafe {
             device.destroy_framebuffer(self.framebuffer);
```

Comment:
We need to be destroying the associated frame buffer *right after* it was presented.
See the quad example for the reference. If keep it, we are forcing Metal to have more than one `CAMetalDrawable` in flight.
Currently, the code appears to be destroying the framebuffer only after the next acquisition is done.